coders/qoi.c

```diff
@@ -0,0 +1,450 @@
+      if (channels == 4)
+        SetPixelAlpha(image,ScaleCharToQuantum((unsigned char)px.rgba.a),q);
+      q+=GetPixelChannels(image);
+      p++;
+    } while (run-- > 0);
+    status=SetImageProgress(image,LoadImageTag,p,image->rows * image->columns);
+    if (status == MagickFalse)
+      break;
+  }
+  if (EOFBlob(image) != MagickFalse)
+    ThrowFileException(exception,CorruptImageError,
+      "UnexpectedEndOfFile",image->filename);
+  if (status != MagickFalse) {
+    status=SyncAuthenticPixels(image,exception);
+    if (image->ping == MagickFalse)
+      SyncImage(image,exception);
+  }
+  (void) CloseBlob(image);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+  return(GetFirstImageInList(image));
+}
+
+
+static MagickBooleanType WriteQOIImage(const ImageInfo *image_info,Image *image,
+  ExceptionInfo *exception)
+{
+  QuantumInfo
+    *quantum_info;
+
+  QuantumType
+    quantum_type;
+
+  MagickBooleanType
+    status;
+
+  const Quantum
+    *p;
+
+  size_t
+    channels,
+    colorspace,
+    end,
+    run,
+    idx,
+    i;
+
+  qoi_rgba_t
+    px,
+    pp,
+    lut[64];
+
+  /*
+    Open output image file.
+  */
+  assert(image_info != (const ImageInfo *) NULL);
+  assert(image_info->signature == MagickCoreSignature);
+  assert(image != (Image *) NULL);
+  assert(image->signature == MagickCoreSignature);
+  if (image->debug != MagickFalse)
+    (void) LogMagickEvent(TraceEvent,GetMagickModule(),"%s",image->filename);
+  assert(exception != (ExceptionInfo *) NULL);
+  assert(exception->signature == MagickCoreSignature);
+  status=OpenBlob(image_info,image,WriteBinaryBlobMode,exception);
+  if (status == MagickFalse)
+    return(status);
+  image->endian=MSBEndian;
+  image->depth=8;
+
+  quantum_info=AcquireQuantumInfo(image_info,image);
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  quantum_type=GetQuantumType(image,exception);
+  if (quantum_type == RGBQuantum)
+    channels=3;
+  else if (quantum_type == RGBAQuantum)
+    channels=4;
+  else
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  if (IssRGBCompatibleColorspace(image->colorspace) == MagickFalse)
+   (void) TransformImageColorspace(image,sRGBColorspace,exception);
+  if (IsRGBColorspace(image->colorspace))
+     colorspace=QOI_LINEAR;
+  else
+     colorspace=QOI_SRGB;
+  /*
+    Write QOI header.
+  */
+  (void) WriteBlobString(image,"qoif");
+  (void) WriteBlobMSBLong(image,(unsigned int) image->columns);
+  (void) WriteBlobMSBLong(image,(unsigned int) image->rows);
+  (void) WriteBlobByte(image,channels);
+  (void) WriteBlobByte(image,colorspace);
+  /*
+    Initialize encoding state.
+  */
+  px.rgba.r=0;
+  px.rgba.g=0;
+  px.rgba.b=0;
+  px.rgba.a=255;
+  memset(lut,0,sizeof(lut));
+  run=0;
+  p=GetVirtualPixels(image,0,0,image->columns,image->rows,exception);
+  if (p == (const Quantum *) NULL)
+    return(MagickFalse);
```

Comment:
Better to do ` ThrowWriterException(ResourceLimitError,"MemoryAllocationFailed");` here instead.

Reply:
done